### ISSUE TITLE
Running a single test

### ISSUE THREAD

**Issue by martotkd1** [bencompton/jest-cucumber#29]
Hello @bencompton ,

Is there a way to run a single test ? 
I am not able to do it by using jest commands.

Thanks,
Martin

**Comment by martotkd1** [comment]
Thank you for the answer, @bencompton . Meanwhile, i will be using vscode-jest.

**Comment by martotkd1** [comment]
In my case , I recently added tags for one feature and there is no filtering at all.

**Comment by Srutherford2407** [comment]
@bencompton any update on this? I am trying to use the tag feature and it appears to work if they all have the tags. But the use case of this is to separate the tests via the tags. 

setup file
`const setJestCucumberConfiguration = require('jest-cucumber').setJestCucumberConfiguration;

setJestCucumberConfiguration({
    tagFilter: ['@test'],
    errorOnMissingScenariosAndSteps: false
});
`
config file
`const { defaults } = require('jest-config');

module.exports = {
  preset: 'jest-puppeteer',
  testPathIgnorePatterns: [...defaults.testPathIgnorePatterns, '.cache'],
  verbose: true,
  setupFiles: ['./qatests/config/jest-cucumber.setup.js'], //Here is the call to the file above
  setupTestFrameworkScriptFile: './qatests/config/jest.setup.js',
  testMatch: [
    "**/qatests/steps/*.steps.js"
  ]
};`

<img width="1298" alt="jesttagerror" src="https://user-images.githubusercontent.com/35584848/48452735-67502000-e77e-11e8-871f-d56fa3990b5f.png">


It looks like it still runs them all and doesnt obey the rule of the tags, but still throws and error like its trying to use the tag rules.

**Comment by bencompton** [comment]
Thanks for the additional info @Srutherford2407. It certainly should not be running scenarios that are excluded via a tag filter. I'll be looking into this soon.

**Comment by Srutherford2407** [comment]
@bencompton Thanks for the update! I would love to give it a test when you implement a fix.

**Comment by bencompton** [comment]
Yeah, sorry for not keeping everyone in the loop.

It is actually published to NPM, at the moment (version 2.0.0). It isn't ready for production yet and is published under the "next" tag instead of "latest", but feel free to try it out at this point.

The new version also supports tag filtering expressions as requested in #32. When defining a tag filter, the tags are no longer specified as an array of strings, and are instead defined as a single string expression. For example, instead of `["ui", "fast"]`, you would use `"ui or fast"`. The API changed, so that necessitated version 2.0.

The "no scenarios found..." errors you were seeing above occur when a scenario is defined in your step definition code, but not in the feature file. In the current build, that error will still occur when you have a tag filter that filters out scenarios that have step definitions implemented. That can be annoying when you're trying to filter out scenarios for a temporary purpose like running only one scenario. My current plan is to make the configuration parameters more granular so it is possible to tell jest-cucumber to ignore the case where there are step definitions that don't have a matching scenario in the feature file, and then make that disabled by default. I think this makes sense to have that disabled by default, since no other implementation of Cucumber bothers to check whether or not there are step definitions that don't match any feature files.

Other than that, I just have to update the docs.

Let me know if you get a chance to try it out and have any feedback.

**Comment by Srutherford2407** [comment]
@bencompton Thanks for the update.

I pulled down @next last night and did some testing. Just wanted to provide feedback from my tests. (I know this is still a work in progress on your side, just wanted to provide my current experience) 

I am using the global configure for the tagFilters, I have a jest-cucumber.config file that I reference from the main jest.config using the Module.exports setupFiles option. Below is the tagFilter for reference.

`setJestCucumberConfiguration({
    tagFilter: ['@Test'],
    errorOnMissingScenariosAndSteps: false
});`

Currently testing with two different feature files (all scenarios in the features have steps created in their respected step files). One feature file contains a tag and the other does not, currently it will run the test containing that tag but throws an error for the other feature file:

`Your test suite must contain at least one test.`

It appears the lack of a matching tag in the second feature file causes the error. If I add the matching tag to that feature file I get the expected result. One test from each feature file runs, and the tests without the tag do not run. Which is awesome!

Console Log for reference: 

` console.log ../node_modules/jest-cucumber/dist/src/tag-filtering.js:32
    (tags.indexOf("@test")!==-1)
  console.log ../node_modules/jest-cucumber/dist/src/tag-filtering.js:51
    [ '@test' ]
  console.log ../node_modules/jest-cucumber/dist/src/tag-filtering.js:53
    true
  console.log ../node_modules/jest-cucumber/dist/src/tag-filtering.js:55
    function anonymous(tags
    ) {
    return (tags.indexOf("@test")!==-1);
    }
  console.log ../node_modules/jest-cucumber/dist/src/tag-filtering.js:51
    []
  console.log ../node_modules/jest-cucumber/dist/src/tag-filtering.js:53
    false
  console.log ../node_modules/jest-cucumber/dist/src/tag-filtering.js:55
    function anonymous(tags
    ) {
    return (tags.indexOf("@test")!==-1);
    }
  console.log ../node_modules/jest-cucumber/dist/src/tag-filtering.js:32
    (tags.indexOf("@test")!==-1)
  console.log ../node_modules/jest-cucumber/dist/src/tag-filtering.js:51
    [ '@test' ]
  console.log ../node_modules/jest-cucumber/dist/src/tag-filtering.js:53`



Also to note as you mentioned in your reply, if I change the config for the error from false to true I still get the error:
No scenarios found in feature file that match scenario title "Test Home Page." (though it does actually exist in my files) But I know you said that would still happen. 


Once again thanks for the update, I am very interested in this feature please reach out if you need anything tested with it!

**Comment by Srutherford2407** [comment]
@bencompton 
Where you able to address this for this issue?

You have two feature files. One feature file contains a tag(@test) and the other does not contain a tag (thus I don't want any tests to run there), currently it will run the test containing that tag but throws an error for the other feature file:

Your test suite must contain at least one test.

Thanks!

**Comment by bencompton** [comment]
My apologies, the last couple of weeks have been hectic, but I am still planning to complete and release this as soon as I have some spare time.

**Comment by bencompton** [comment]
I published v2.0.1 to address the issues with tag filtering. Previously, scenarios that didn't match the tag were not getting registered with Jest, which was causing Jest to throw the error about the test suite needing to contain at least one test. With this update, jest-cucumber will now mark tests as skipped instead of simply excluding them.

This is currently published to the beta channel ("next" tag). Once I do more exhaustive testing and get more feedback from the community, I'll be moving this to the stable channel.

**Comment by bencompton** [comment]
@srutherford2407 - thanks for having a look. The tag filter now takes a single string expression instead of an array. You can string together multiple tags with ands and ors.

**Comment by Srutherford2407** [comment]
@bencompton  

Thanks! I just had to adjust the syntax but this code does work and ran all the correct tests.

```
setJestCucumberConfiguration({
    tagFilter: " '@Test' and '@Test2' ", 
    errorOnMissingScenariosAndSteps: false
});
```

**Comment by bencompton** [comment]
@srutherford2407 - Glad it worked for you. You actually shouldn’t need the single quotes in the tag filter. This should work: ‘@Test and @Test2’.

**Comment by Srutherford2407** [comment]
@bencompton 
I tired with that syntax originally, but written like this it failed which its why I tried the additional quotes (which did work)
```
setJestCucumberConfiguration({
    tagFilter: '@Test and @Test2', 
    errorOnMissingScenariosAndSteps: false
});
```

This also runs everything I expected it to run:
```
setJestCucumberConfiguration({
    tagFilter: " '@Test and  @Test2' ", 
    errorOnMissingScenariosAndSteps: false
});
```

**Comment by martotkd1** [comment]
Hello guys,
Update from me as well.
I upgraded to v2.0.2 and ran all of my tests and everything works fine. No regressions.
Tagging also works fine.
The only thing that bottering me is, if i want to run the tests based on config, should i always tag all the tests in the default scenario ? And then add additional tags for scenarios that will be run for particular config ?

**Comment by bencompton** [comment]
@martotkd1, not sure I follow what you mean. Perhaps you could elaborate on what you're looking to accomplish?

**Comment by martotkd1** [comment]
I figured it out, my typo.
Everything works fine on my side.

**Comment by martotkd1** [comment]
Just a funny update on my side.
Ironically i wanted to use the following tags - @sandbox @ignore - and filtering does not work because they contains 'and' and 'or' :) :).

**Comment by martotkd1** [comment]
One more thing.
The doc says that tagFilter: '@included and not @excluded' will run the scenario tagged @included, and will skip the scenario tagged @excluded. 

In my case all the tests except @included are skipped.

If i type @included or not @excluded, then @included is run along all other scenarios except @excluded.

**Comment by bencompton** [comment]
@martotkd1 and @Srutherford2407, thanks for testing and reporting issues! Sorry it has taken me so long to get around to addressing these issues, but I just pushed out v2.0.3 with some fixes:

1. @Srutherford2407's issue where `" '@Test and  @Test2' "` works and `'@Test and  @Test2'` doesn't was occurring because the extra quotes was making it get evaluated as a string instead of a boolean expression. Since a non-empty string evaluates to true in JavaScript, the tag expression always gets evaluated to true. I fixed it so that extra quotes result in an error parsing the tag expression. Also, I was only able to reproduce the incorrect behavior by tagging a scenario with @Test Test2 (omitting the @ on Test2). I am guessing you might have done something like that, because I wasn't able to reproduce your original issue any other way.

2. @martotkd1, your issue where @sandbox gets interpreted incorrectly because it contains "and" should now be fixed (had to refine some regular expressions for recognizing the keywords `and`, `or`, and `not`).
